src/Covenant/Prim.hs

```diff
@@ -279,3 +295,128 @@ data SixArgFunc
+  BLS12_381_G1_add -> (TyByteString, TyByteString, TyByteString)
+  BLS12_381_G1_scalarMul -> (TyInteger, TyByteString, TyByteString)
+  BLS12_381_G1_equal -> (TyByteString, TyByteString, TyBoolean)
+  BLS12_381_G1_hashToGroup -> (TyByteString, TyByteString, TyByteString)
+  BLS12_381_G2_add -> (TyByteString, TyByteString, TyByteString)
+  BLS12_381_G2_scalarMul -> (TyInteger, TyByteString, TyByteString)
+  BLS12_381_G2_equal -> (TyByteString, TyByteString, TyBoolean)
+  BLS12_381_G2_hashToGroup -> (TyByteString, TyByteString, TyByteString)
+  BLS12_381_millerLoop -> (TyByteString, TyByteString, TyByteString)
+  BLS12_381_mulMlResult -> (TyByteString, TyByteString, TyByteString)
+  BLS12_381_finalVerify -> (TyByteString, TyByteString, TyBoolean)
+  ByteStringToInteger -> (TyBoolean, TyByteString, TyInteger)
+  ReadBit -> (TyByteString, TyInteger, TyBoolean)
+  ReplicateByte -> (TyInteger, TyInteger, TyByteString)
+  ShiftByteString -> (TyByteString, TyInteger, TyByteString)
+  RotateByteString -> (TyByteString, TyInteger, TyByteString)
+
+-- | Maps three-argument functions to their input and output types.
+-- Returns a tuple of (arg1 type, arg2 type, arg3 type, return type).
+--
+-- @since 1.0.0
+typeOfThreeArgFunc :: ThreeArgFunc -> (TyConstant, TyConstant, TyConstant, TyConstant)
+typeOfThreeArgFunc = \case
+  VerifyEd25519Signature -> (TyByteString, TyByteString, TyByteString, TyBoolean)
+  VerifyEcdsaSecp256k1Signature -> (TyByteString, TyByteString, TyByteString, TyBoolean)
+  VerifySchnorrSecp256k1Signature -> (TyByteString, TyByteString, TyByteString, TyBoolean)
+  IfThenElse -> (TyBoolean, TyPlutusData, TyPlutusData, TyPlutusData)
+  ChooseList -> (TyList TyPlutusData, TyPlutusData, TyPlutusData, TyPlutusData)
+  CaseList -> (TyList TyPlutusData, TyPlutusData, TyPair TyPlutusData (TyList TyPlutusData), TyPlutusData)
+  IntegerToByteString -> (TyBoolean, TyInteger, TyInteger, TyByteString)
+  AndByteString -> (TyBoolean, TyByteString, TyByteString, TyByteString)
+  OrByteString -> (TyBoolean, TyByteString, TyByteString, TyByteString)
+  XorByteString -> (TyBoolean, TyByteString, TyByteString, TyByteString)
+  WriteBits -> (TyByteString, TyList TyInteger, TyBoolean, TyByteString)
+  ExpModInteger -> (TyInteger, TyInteger, TyInteger, TyInteger)
+
+-- | Maps six-argument functions to their input and output types.
+-- Returns a tuple of (arg1 type, arg2 type, arg3 type, .. arg6 type, return type).
+--
+-- @since 1.0.0
+typeOfSixArgFunc :: SixArgFunc -> (TyConstant, TyConstant, TyConstant, TyConstant, TyConstant, TyConstant, TyConstant)
+typeOfSixArgFunc = \case
+  ChooseData -> (TyPlutusData, TyPlutusData, TyPlutusData, TyPlutusData, TyPlutusData, TyPlutusData, TyPlutusData)
+  CaseData ->
+    ( TyPlutusData,
+      TyPair TyInteger (TyList TyPlutusData),
+      TyList (TyPair TyPlutusData TyPlutusData),
+      TyList TyPlutusData,
+      TyInteger,
+      TyByteString,
+      TyPlutusData
+    )
```

Comment:
If you look at the [stated semantics of `CaseData`](https://github.com/IntersectMBO/plutus/blob/master/plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs#L2035-L2050), you will see that it doesn't take arbitrary polymorphic terms, but instead functions with a polymorphic return type. I assume that this is just 'shorthanded' the same way as other polymorphic functions?

Reply:
ooh right I meant to update this once I figure out the function types in AConstant. Thanks for catching this.
For some reason I thought PrimCall can only accept constants.
These were not shorthanded, I'll replace these with TyLam.

Reply:
`CaseData` is a somewhat unusual situation, as it's the _only_ builtin that does this.